packages/core/src/config/parse.ts

```diff
@@ -1722,41 +1730,97 @@ const assertValidConstructorArgs = (
   // We return the cached values so we can use them in later steps without rereading the files
   return {
     cachedCompilerOutput,
-    cachedConstructorArgs,
-    cachedArtifacts,
+    parsedConstructorArgs,
   }
 }
 
-const resolveContractAddresses = (
+export const resolveContractAddresses = async (
   userConfig: UserChugSplashConfig,
-  cachedArtifacts: { [referenceName: string]: ContractArtifact }
-): { [referenceName: string]: string } => {
+  artifactPaths: ArtifactPaths,
+  integration: Integration,
+  stream: NodeJS.WritableStream,
+  silent: boolean
+): Promise<{ [referenceName: string]: string }> => {
   // Resolve all the contract addresses so they can be used handle contract references regardless or ordering.
   const contracts: { [referenceName: string]: string } = {}
   for (const [referenceName, userContractConfig] of Object.entries(
     userConfig.contracts
   )) {
     const { externalProxy, kind } = userContractConfig
 
-    // Set the proxy address to the user-defined value if it exists, otherwise set it to the default proxy
-    // used by ChugSplash.
-    const proxy =
-      externalProxy ||
-      getDefaultProxyAddress(
-        userConfig.options.organizationID,
-        userConfig.options.projectName,
-        referenceName
+    // if the contract is proxied then we can just return the proxy address
+    if (kind !== 'no-proxy') {
+      // Set the proxy address to the user-defined value if it exists, otherwise set it to the default proxy
+      // used by ChugSplash.
+      const proxy =
+        externalProxy ||
+        getDefaultProxyAddress(
+          userConfig.options.organizationID,
+          userConfig.options.projectName,
+          referenceName
+        )
+
+      contracts[referenceName] = proxy
+    } else {
+      // if the contract is not proxied then we need to resolve the actual contract address
+
+      // If non-proxied contract includes libraries, then we must resolve any references used when defining those libraries.
+      // This is because library addresses are injected into the bytecode which will impact the create2 address of the contract
+      // which we are attempting to resolve here.
+
+      // So we first compile the libraries using the contract addresses we have already resolved.
+      // This allows users to use contract references to define libraries in non-proxied contracts.
+      // Unfortunately, this means that users must define the library before it is used or this process will fail.
+      const compiledLibraries: Libraries = JSON.parse(
+        Handlebars.compile(JSON.stringify(userContractConfig.libraries ?? {}))({
+          ...contracts,
+        })
```

Comment:
> Could we wrap the library logic in an if-statement that's only entered if userContractConfig.libraries is defined and has more than one field?
It would need to be if the number of libraries is greater than 0, but yeah we can do that basically. We still need this if there is only one entry in `userContractConfig.libraries` because this logic is used to resolve dependencies between multiple no-proxy contracts. I.e A -> B -> C is a case where each contract has only one library, but they need to be correctly ordered if the contracts are not proxied.